ops/charm.py

```diff
@@ -823,77 +826,105 @@ def defer(self):
 class CharmEvents(ObjectEvents):
     """Events generated by Juju pertaining to application lifecycle.
 
-    This class is used to create an event descriptor (``self.on``) attribute for
+    This class is used to create an event descriptor (``self.on``) for
     a charm class that inherits from :class:`CharmBase`. The event descriptor
     may be used to set up event handlers for corresponding events.
 
-    By default the following events will be provided through
-    :class:`CharmBase`::
-
-        self.on.install
-        self.on.start
-        self.on.remove
-        self.on.update_status
-        self.on.config_changed
-        self.on.upgrade_charm
-        self.on.pre_series_upgrade
-        self.on.post_series_upgrade
-        self.on.leader_elected
-        self.on.collect_metrics
-
-
-    In addition to these, depending on the charm's metadata (``metadata.yaml``),
-    named relation and storage events may also be defined.  These named events
-    are created by :class:`CharmBase` using charm metadata.  The named events may be
-    accessed as ``self.on[<name>].<relation_or_storage_event>``
+    By default, the events listed below will be provided via the
+    :attr:`CharmBase.on` attribute. For example::
+
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    In addition to the events listed below, dynamically-named events will also
+    be defined based on the charm's metadata (``metadata.yaml``) for relations,
+    storage, actions, and containers. These named events may be accessed as
+    ``self.on[<name>].<event>`` or using a prefix like ``self.on.<name>_<event>``,
+    for example::
+
+        self.framework.observe(self.on["db"].relation_created, self._on_db_relation_created)
+        self.framework.observe(self.on.workload_pebble_ready, self._on_workload_pebble_ready)
     """
 
+    # NOTE: The one-line docstrings below are copied from the first line of
+    #       each event class's docstring. Please keep in sync.
```

Comment:
could do that programmatically?

Reply:
some `inspect` magic...

Reply:
I had a play with this, and unfortunately these attribute docstrings aren't inspectable. They're a Sphinx / IDE feature and don't seem to correspond to a `__doc__` attribute. So I'm going to leave this as is for now.